File: services/servico-order/src/infrastructure/repository/InMemoryOrderRepository.ts
import { Order } from '../../domain/Order';
import { OrderRepository } from '../../application/ports/OrderRepository';

/**
 * Implementação em memória da porta OrderRepository. Adequada para
 * demonstração, testes automatizados e desenvolvimento local. A troca
 * por um adaptador PostgreSQL (orders_db, ver EC-M4.3/EC-M5.2) não
 * exige nenhuma mudança na camada de aplicação ou de domínio.
 */
export class InMemoryOrderRepository implements OrderRepository {
  private readonly store = new Map<string, Order>();

  async save(order: Order): Promise<void> {
    this.store.set(order.id, order);
  }

  async findById(id: string): Promise<Order | null> {
    return this.store.get(id) ?? null;
  }

  /** Utilitário de teste/depuração — lista todos os pedidos persistidos. */
  async findAll(): Promise<Order[]> {
    return Array.from(this.store.values());
  }

  async clear(): Promise<void> {
    this.store.clear();
  }
}

File: services-java/servico-order/src/main/java/br/com/ftgo/order/infrastructure/persistence/InMemoryOrderRepository.java
package br.com.ftgo.order.infrastructure.persistence;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.stereotype.Repository;

import br.com.ftgo.order.application.port.OrderRepository;
import br.com.ftgo.order.domain.Order;

/**
 * Implementação em memória da porta {@link OrderRepository}. Adequada
 * para demonstração, testes automatizados e desenvolvimento local. A
 * troca por um adaptador PostgreSQL/JPA ({@code orders_db}, ver
 * EC-M4.3/EC-M5.2) não exige nenhuma mudança na camada de aplicação ou
 * de domínio.
 */
@Repository
public class InMemoryOrderRepository implements OrderRepository {

    private final ConcurrentHashMap<String, Order> store = new ConcurrentHashMap<>();

    @Override
    public void save(Order order) {
        store.put(order.getId(), order);
    }

    @Override
    public Optional<Order> findById(String id) {
        return Optional.ofNullable(store.get(id));
    }

    @Override
    public List<Order> findAll() {
        return List.copyOf(store.values());
    }
}
